perf(db): Reuse cursors in the Latest/HistoricalStateProviders

mediocregopher opened the pull request:
This PR introduces a `ReusableCursor` type which automatically returns a cursor to a Cell on Drop. By holding a Cell for each table type we are able to reuse cursors in the state providers.

I wanted to make this even more generic and automatic, so that we automatically reuse dropped cursors on every Tx, but doing so is a bigger lift. This should at least get us something useful.

Closes #20492


Update:

This introduces a owned/borrowed enum so that we can propagate borrowed cursors and instantiate owned resuable cursors, which we currently do in here:

https://github.com/paradigmxyz/reth/blob/646fec9b4b36707b48d8271715c69d9c38cfddf8/crates/storage/provider/src/providers/database/provider.rs#L203-L203

we could further add an instance to:

https://github.com/paradigmxyz/reth/blob/646fec9b4b36707b48d8271715c69d9c38cfddf8/crates/storage/provider/src/providers/database/provider.rs#L154-L154

so that we can borrow there as well.

this is currently used in the execution stage:

https://github.com/paradigmxyz/reth/blob/646fec9b4b36707b48d8271715c69d9c38cfddf8/crates/stages/stages/src/stages/execution.rs#L298-L298

which is used for multi block execution so I believe the owned variant would be beneficial here as well

Closes RETH-179

mediocregopher committed:
reth-provider compiles

mediocregopher committed:
WIP

mediocregopher committed:
InMemoryOverlayProvider

mediocregopher committed:
WIP

mediocregopher committed:
Compiling

mediocregopher committed:
Got rid of extra state provider in payload building

mediocregopher committed:
Final clippy fixes

mediocregopher committed:
doc fix

mediocregopher committed:
WIP

mediocregopher committed:
use an option rather than ManuallyDrop

mediocregopher committed:
remove tx lifetime

mediocregopher committed:
elided lifetimes

mediocregopher committed:
clippy

mediocregopher committed:
Merge remote-tracking branch 'upstream/main' into mediocregopher/historical-provider-reuse-cursors

mattsse reviewed (approved):
this makes sense,

but I believe all provider trait fn go through

https://github.com/paradigmxyz/reth/blob/baa97b38686a6d63d3d16b2620bbfcb73075420d/crates/storage/storage-api/src/macros.rs#L1-L21

which rn would create a new instance of the reusablecursors per as_ref call
so this only gives a benefit if a function of

https://github.com/paradigmxyz/reth/blob/baa97b38686a6d63d3d16b2620bbfcb73075420d/crates/storage/storage-api/src/macros.rs#L27-L31

would create two cursors for the same table

mattsse committed:
Merge branch 'main' into mediocregopher/historical-provider-reuse-cursors

mattsse committed:
feat: add cursor cache wrapper

mattsse committed:
Merge branch 'main' into mediocregopher/historical-provider-reuse-cursors

decofe committed:
Merge main into mediocregopher/historical-provider-reuse-cursors

Resolved conflicts by taking main's EitherReader approach for history
lookups. The cursor reuse changes are incompatible with the new
multi-backend (MDBX/RocksDB) architecture introduced in main.

Amp-Thread-ID: https://ampcode.com/threads/T-019c009c-a902-730e-b0a9-ae531f17b324
Co-authored-by: Amp <amp@ampcode.com>

decofe committed:
fix: remove unused cursor reuse fields and make as_ref const

- Remove unused fields (storage_changesets, accounts_history, storages_history)
  from ReusableStateCursors since only latest.rs uses cursor reuse now
- Remove corresponding unused methods
- Make HistoricalStateProvider::as_ref and new_with_lowest_available_blocks const

Amp-Thread-ID: https://ampcode.com/threads/T-019c009c-a902-730e-b0a9-ae531f17b324
Co-authored-by: Amp <amp@ampcode.com>

gakonst committed:
Merge remote-tracking branch 'origin/main' into mediocregopher/historical-provider-reuse-cursors

DaniPopes commented on crates/storage/provider/src/providers/state/cursor_reuse.rs:
```diff
@@ -0,0 +1,91 @@
+///
+/// This reduces cursor allocation overhead for state providers that perform many sequential
+/// database operations.
+pub(crate) struct ReusableStateCursors<TX: DbTx> {
```
whys this plural
